src/map_header_data.c

```diff
@@ -0,0 +1,281 @@
+#include <nitro.h>
+#include <string.h>
+
+#include "field/field_system.h"
+#include "overlay006/struct_ov6_02242634.h"
+
+#include "narc.h"
+#include "heap.h"
+#include "map_header.h"
+#include "map_header_data.h"
+#include "map_object.h"
+
+#include "constants/heap.h"
+
+static void MapHeaderData_LoadEvents(MapHeaderData * data, int headerID);
+static void MapHeaderData_ParseEvents(MapHeaderData * data);
+static void MapHeaderData_LoadScripts(MapHeaderData * data, int headerID);
+
+void MapHeaderData_Init (FieldSystem * fieldSystem, enum HeapId heapID)
+{
+    GF_ASSERT(fieldSystem->mapHeaderData == NULL);
+    fieldSystem->mapHeaderData = Heap_AllocFromHeap(heapID, sizeof(MapHeaderData));
+}
+
+void MapHeaderData_Free (FieldSystem * fieldSystem)
+{
+    GF_ASSERT(fieldSystem->mapHeaderData != NULL);
+    Heap_FreeToHeap(fieldSystem->mapHeaderData);
+}
+
+void MapHeaderData_Load (FieldSystem * fieldSystem, int headerID)
+{
+    GF_ASSERT(fieldSystem->mapHeaderData != NULL);
+
+    MapHeaderData_LoadEvents(fieldSystem->mapHeaderData, headerID);
+    MapHeaderData_ParseEvents(fieldSystem->mapHeaderData);
+    MapHeaderData_LoadWildEncounters(&fieldSystem->mapHeaderData->wildEncounters, headerID);
+    MapHeaderData_LoadScripts(fieldSystem->mapHeaderData, headerID);
+}
+
+static void MapHeaderData_LoadEvents (MapHeaderData * data, int headerID)
+{
+    int eventsID = MapHeader_GetEventsArchiveID(headerID);
+    GF_ASSERT(NARC_GetMemberSizeByIndexPair(NARC_INDEX_FIELDDATA__EVENTDATA__ZONE_EVENT, eventsID) < sizeof(data->tmpEventsBuf));
+    NARC_ReadWholeMemberByIndexPair(data->tmpEventsBuf, NARC_INDEX_FIELDDATA__EVENTDATA__ZONE_EVENT, eventsID);
+}
+
+void sub_0203A418 (FieldSystem * fieldSystem)
+{
+    int numObjectEvents = fieldSystem->mapHeaderData->numObjectEvents;
+
+    GF_ASSERT(fieldSystem->mapHeaderData != NULL);
+
+    if (numObjectEvents != 0) {
+        sub_02062068(fieldSystem->mapObjMan, fieldSystem->location->mapId, numObjectEvents, fieldSystem->mapHeaderData->objectEvents);
+    }
+}
+
+const BgEvent * MapHeaderData_GetBgEvents (const FieldSystem * fieldSystem)
+{
+    return fieldSystem->mapHeaderData->bgEvents;
+}
+
+int MapHeaderData_GetNumBgEvents (const FieldSystem * fieldSystem)
+{
+    return fieldSystem->mapHeaderData->numBgEvents;
+}
+
+const WarpEvent * MapHeaderData_GetWarpEventByIndex (const FieldSystem * fieldSystem, int index)
+{
+    return (index >= fieldSystem->mapHeaderData->numWarpEvents)
+        ? NULL
+        : &fieldSystem->mapHeaderData->warpEvents[index];
+}
+
+int MapHeaderData_GetIndexOfWarpEventAtPos (const FieldSystem * fieldSystem, int x, int z)
+{
+    for (int i = 0; i < fieldSystem->mapHeaderData->numWarpEvents; i++) {
+        if (fieldSystem->mapHeaderData->warpEvents[i].x == x
+            && fieldSystem->mapHeaderData->warpEvents[i].z == z) {
```

Comment:
Maybe macro-able? There's at least one other place in this file with a similar construct. I'm thinking, loosely:

```c
#define AT_COORDS(event, xPos, zPos) event.x == xPos && event.z == zPos
```

Reply:
I don't see the point of adding another layer of indirection? The code is imo already pretty readable to me and this pattern occurs only one other time.